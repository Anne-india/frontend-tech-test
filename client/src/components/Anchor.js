import React from 'react';
import { Link } from 'react-router-dom';

const Anchor = props => {
    const {className, to, text} = props;
    return (
        <Link className={className} to={to}>{text}</Link>
    );
};

export default Anchor;