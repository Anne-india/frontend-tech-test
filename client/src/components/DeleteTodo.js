import React from 'react';

const DeleteTodo = props => {
    return (
        <button className="btn" onClick={() => props.deleteTodo(props.id)}>
            <i className="fa fa-trash"></i>
        </button>
    );
};

export default DeleteTodo;


