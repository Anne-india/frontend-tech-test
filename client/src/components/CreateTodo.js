import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { createTodo } from '../actions';
import Button from './Button';
import Anchor from './Anchor';

export class CreateTodo extends Component {
    handleCreate = e => {
        e.preventDefault();
        this.props.createTodo(this.refs.title.value, this.refs.description.value);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <form onSubmit={this.handleCreate}>
                    <h4 className="other-title">Create new task</h4>
                    <div className="form-group">
                        <label htmlFor="title">Title</label>
                        <input type="text" className="form-control" placeholder="Give a title..." required
                            name="title" ref="title" />
                    </div>
                    <div className="form-group">
                        <label htmlFor="description">Description</label>
                        <input type="text" className="form-control" placeholder="Add a description..." required
                            name="description" ref="description" />
                    </div>
                    <div className="text-center">
                        <div className="btn-group" role="group">
                            <Anchor className="btn btn-light mr-4" to="/" text="Back"/>
                            <Button type="submit" className="btn btn-primary" text="Create"/>
                        </div>
                    </div>
                </form >
            </div >
        );
    }
}

CreateTodo.propTypes = {
    createTodo: PropTypes.func.isRequired
};

const mapDispatchToProps = {
    createTodo
}

export default connect(null, mapDispatchToProps)(CreateTodo);