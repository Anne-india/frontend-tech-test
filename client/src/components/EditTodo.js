import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchIndividualTodo, updateTodo } from '../actions';
import Button from './Button';
import Anchor from './Anchor';

export class EditTodo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            id: '',
            title: '',
            description: ''
        }
    }

    handleEdit = e => {
        e.preventDefault();
        this.props.updateTodo(this.state.id, this.state.title, this.state.description);
        this.props.history.push('/');
    }
    handleChange = e => {
        this.setState({
            [e.target.name]: e.target.value
        })
    }

    componentWillMount() {
        this.props.fetchIndividualTodo(this.props.match.params.id);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            id: nextProps.getTodoTask.id,
            title: nextProps.getTodoTask.title,
            description: nextProps.getTodoTask.description
        })
    }

    render() {
        return (
            <div>
                <form onSubmit={this.handleEdit}>
                    <input type="hidden" required
                        name="id" value={this.state.id || ''} />
                    <h4 className="other-title">Update task</h4>
                    <div className="form-group">
                        <label htmlFor="title">Title</label>
                        <input type="text" className="form-control" required
                            name="title" ref="title" value={this.state.title || ''}
                            onChange={e => this.handleChange(e)} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="description">Description</label>
                        <input type="text" className="form-control" required
                            name="description" ref="description" value={this.state.description || ''}
                            onChange={e => this.handleChange(e)} />
                    </div>
                    <div className="text-center">
                        <div className="btn-group" role="group">
                            <Anchor className="btn btn-light mr-4" to="/" text="Back"/>
                            <Button type="submit" className="btn btn-primary" text="Update"/>
                        </div>
                    </div>
                </form >
            </div >
        );
    }
}

EditTodo.propTypes = {
    fetchIndividualTodo: PropTypes.func.isRequired,
    updateTodo: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {
        getTodoTask: state.getTodoList.task
    }
}

const mapDispatchToProps = {
    fetchIndividualTodo,
    updateTodo
}

export default connect(mapStateToProps, mapDispatchToProps)(EditTodo);