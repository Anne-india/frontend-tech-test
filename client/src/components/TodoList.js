import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { fetchTodoList, deleteTodo } from '../actions';
import DeleteTodo from './DeleteTodo';

export class TodoList extends Component {
    componentWillMount() {
        this.props.fetchTodoList();
    }

    handleDelete = id => {
        this.props.deleteTodo(id);
    }

    //generate todo list
    createList = getTodoList => {
        return getTodoList.map(todo => {
            return (
                <div className="item" key={todo.id}>
                    <div className="details">
                        <div className="title">
                            {todo.title}
                        </div>
                        <div className="description">
                            {todo.description}
                        </div>
                    </div>
                    <div className="actions">
                        <div className="edit">
                            <Link className="btn" to={`/edit/${todo.id}`}>
                                <i className="fa fa-pencil" aria-hidden="true"></i>
                            </Link>
                        </div>
                        <div className="delete">
                            <DeleteTodo id={todo.id} deleteTodo={this.handleDelete} />
                        </div>
                    </div>
                </div>
            )
        })
    }
    /* show alert when no tasks are created/found */
    showEmptyMessage = () => {
        return (
            <p className="alert alert-primary custom-alert" role="alert">No tasks found! Please create a new task</p>
        )
    }

    renderList = () => {
        if (this.props.getTodoList.length !== 0) {
            return this.createList(this.props.getTodoList);
        }
        return this.showEmptyMessage();
    }

    render() {
        return (
            <Fragment>
                {this.renderList()}
                <div className="actions create">
                    <Link className="btn" to='/create'>
                        <i className="fa fa-plus" aria-hidden="true"></i>
                    </Link>
                </div>
            </Fragment>
        );
    }
}

TodoList.propTypes = {
    fetchTodoList: PropTypes.func.isRequired,
    deleteTodo: PropTypes.func.isRequired,
    getTodoList: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        title: PropTypes.string,
        description: PropTypes.string
    }))
};

const mapStateToProps = (state) => {
    return {
        getTodoList: state.getTodoList.tasks
    };
}

export default connect(mapStateToProps, { fetchTodoList, deleteTodo })(TodoList);