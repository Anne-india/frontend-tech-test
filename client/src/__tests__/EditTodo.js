import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { EditTodo } from '../components/EditTodo';
import renderer from 'react-test-renderer';

describe('Edit Component', () => {
  let props;

  beforeEach(() => {
    props = {
        fetchIndividualTodo : jest.fn(),
        updateTodo : jest.fn(),
    };
  });

  it('renders without crashing', () => {
    const tree = renderer.create(
        <BrowserRouter>
        <EditTodo {...props} match={{params: {id: 1}, isExact: true, path: "", url: ""}}/>
        </BrowserRouter>)
    expect(tree.toJSON()).toMatchSnapshot();
  });
});