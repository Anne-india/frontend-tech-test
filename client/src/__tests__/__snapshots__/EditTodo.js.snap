// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component renders without crashing 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      name="id"
      required={true}
      type="hidden"
      value=""
    />
    <h4
      className="other-title"
    >
      Update task
    </h4>
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        className="form-control"
        name="title"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="description"
      >
        Description
      </label>
      <input
        className="form-control"
        name="description"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="text-center"
    >
      <div
        className="btn-group"
        role="group"
      >
        <a
          className="btn btn-light mr-4"
          href="/"
          onClick={[Function]}
        >
          Back
        </a>
        <button
          className="btn btn-primary"
          type="submit"
        >
          Update
        </button>
      </div>
    </div>
  </form>
</div>
`;
